// typedoc/dist/lib/output/plugins/NavigationPlugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationPlugin = void 0;
const components_1 = require("../components");
const events_1 = require("../events");
/**
 * A plugin that exposes the navigation structure of the documentation
 * to the rendered templates.
 *
 * The navigation structure is generated using the current themes
 * [[BaseTheme.getNavigation]] function. This plugins takes care that the navigation
 * is updated and passed to the render context.
 */
let NavigationPlugin = class NavigationPlugin extends components_1.RendererComponent {
    /**
     * Create a new NavigationPlugin instance.
     */
    initialize() {
        this.listenTo(this.owner, {
            [events_1.RendererEvent.BEGIN]: this.onBeginRenderer,
            [events_1.PageEvent.BEGIN]: this.onBeginPage,
            [events_1.RendererEvent.END]: () => (this.navigation = void 0),
        });
    }
    /**
     * Triggered before the renderer starts rendering a project.
     *
     * @param event  An event object describing the current render operation.
     */
    onBeginRenderer(event) {
        this.navigation = this.owner.theme.getNavigation(event.project);
    }
    /**
     * Triggered before a document will be rendered.
     *
     * @param page  An event object describing the current render operation.
     */
    onBeginPage(page) {
        const currentItems = [];
        (function updateItem(item) {
            item.isCurrent = false;
            item.isInPath = false;
            item.isVisible = item.isModules;
            if (item.url === page.url ||
                (item.dedicatedUrls && item.dedicatedUrls.includes(page.url))) {
                currentItems.push(item);
            }
            if (item.children) {
                item.children.forEach((child) => updateItem(child));
            }
        })(this.navigation);
        currentItems.forEach((item) => {
            item.isCurrent = true;
            let depth = item.isModules ? -1 : 0;
            let count = 1;
            while (item) {
                item.isInPath = true;
                item.isVisible = true;
                count += 1;
                depth += 1;
                if (item.children) {
                    count += item.children.length;
                    if (depth < 2 || count < 30) {
                        item.children.forEach((child) => {
                            child.isVisible = true;
                        });
                    }
                }
                item = item.parent;
            }
        });
        page.navigation = this.navigation;
    }
};
NavigationPlugin = __decorate([
    (0, components_1.Component)({ name: "navigation" })
], NavigationPlugin);
exports.NavigationPlugin = NavigationPlugin;
