// typedoc/dist/lib/output/plugins/PrettyPrintPlugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var PrettyPrintPlugin_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrettyPrintPlugin = void 0;
const components_1 = require("../components");
const events_1 = require("../events");
/**
 * List of states the parser of [[PrettyPrintPlugin]] can be in.
 */
var PrettyPrintState;
(function (PrettyPrintState) {
    /**
     * Default state of the parser. Empty lines will be removed and indention will be adjusted.
     */
    PrettyPrintState[PrettyPrintState["Default"] = 0] = "Default";
    /**
     * Comment state, the parser waits for a comment closing tag.
     */
    PrettyPrintState[PrettyPrintState["Comment"] = 1] = "Comment";
    /**
     * Pre state, the parser waits for the closing tag of the current pre block.
     */
    PrettyPrintState[PrettyPrintState["Pre"] = 2] = "Pre";
})(PrettyPrintState || (PrettyPrintState = {}));
/**
 * A plugin that pretty prints the generated html.
 *
 * This not only aids in making the generated html source code more readable, by removing
 * blank lines and unnecessary whitespaces the size of the documentation is reduced without
 * visual impact.
 *
 * At the point writing this the docs of TypeDoc took 97.8 MB  without and 66.4 MB with this
 * plugin enabled, so it reduced the size to 68% of the original output.
 */
let PrettyPrintPlugin = PrettyPrintPlugin_1 = class PrettyPrintPlugin extends components_1.RendererComponent {
    /**
     * Create a new PrettyPrintPlugin instance.
     */
    initialize() {
        this.listenTo(this.owner, events_1.PageEvent.END, this.onRendererEndPage, -1024);
    }
    /**
     * Triggered after a document has been rendered, just before it is written to disc.
     *
     * @param event
     */
    onRendererEndPage(event) {
        let match;
        let line;
        let lineState;
        let lineDepth;
        let tagName;
        let preName;
        const tagExp = /<\s*(\w+)[^>]*>|<\/\s*(\w+)[^>]*>|<!--|-->/g;
        const emptyLineExp = /^[\s]*$/;
        let minLineDepth = 1;
        let state = PrettyPrintState.Default;
        const stack = [];
        const lines = (event.contents || "").split(/\r\n?|\n/);
        let index = 0;
        let count = lines.length;
        while (index < count) {
            line = lines[index];
            if (emptyLineExp.test(line)) {
                if (state === PrettyPrintState.Default) {
                    lines.splice(index, 1);
                    count -= 1;
                    continue;
                }
            }
            else {
                lineState = state;
                lineDepth = stack.length;
                while ((match = tagExp.exec(line))) {
                    if (state === PrettyPrintState.Comment) {
                        if (match[0] === "-->") {
                            state = PrettyPrintState.Default;
                        }
                    }
                    else if (state === PrettyPrintState.Pre) {
                        if (match[2] && match[2].toLowerCase() === preName) {
                            state = PrettyPrintState.Default;
                        }
                    }
                    else {
                        if (match[0] === "<!--") {
                            state = PrettyPrintState.Comment;
                        }
                        else if (match[1]) {
                            tagName = match[1].toLowerCase();
                            if (tagName in PrettyPrintPlugin_1.IGNORED_TAGS) {
                                continue;
                            }
                            if (tagName in PrettyPrintPlugin_1.PRE_TAGS) {
                                state = PrettyPrintState.Pre;
                                preName = tagName;
                            }
                            else {
                                if (tagName === "body") {
                                    minLineDepth = 2;
                                }
                                stack.push(tagName);
                            }
                        }
                        else if (match[2]) {
                            tagName = match[2].toLowerCase();
                            if (tagName in PrettyPrintPlugin_1.IGNORED_TAGS) {
                                continue;
                            }
                            const n = stack.lastIndexOf(tagName);
                            if (n !== -1) {
                                stack.length = n;
                            }
                        }
                    }
                }
                if (lineState === PrettyPrintState.Default) {
                    lineDepth = Math.min(lineDepth, stack.length);
                    line = line.replace(/^\s+/, "").replace(/\s+$/, "");
                    if (lineDepth > minLineDepth) {
                        line =
                            Array(lineDepth - minLineDepth + 1).join("\t") +
                                line;
                    }
                    lines[index] = line;
                }
            }
            index++;
        }
        event.contents = lines.join("\n");
    }
};
/**
 * Map of all tags that will be ignored.
 */
PrettyPrintPlugin.IGNORED_TAGS = {
    area: true,
    base: true,
    br: true,
    wbr: true,
    col: true,
    command: true,
    embed: true,
    hr: true,
    img: true,
    input: true,
    link: true,
    meta: true,
    param: true,
    source: true,
};
/**
 * Map of all tags that prevent this plugin form modifying the following code.
 */
PrettyPrintPlugin.PRE_TAGS = {
    pre: true,
    code: true,
    textarea: true,
    script: true,
    style: true,
};
PrettyPrintPlugin = PrettyPrintPlugin_1 = __decorate([
    (0, components_1.Component)({ name: "pretty-print" })
], PrettyPrintPlugin);
exports.PrettyPrintPlugin = PrettyPrintPlugin;
