// typedoc/dist/lib/models/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownType = exports.UnionType = exports.TypeParameterType = exports.TypeOperatorType = exports.TupleType = exports.NamedTupleMember = exports.TemplateLiteralType = exports.RestType = exports.ReflectionType = exports.ReferenceType = exports.QueryType = exports.PredicateType = exports.OptionalType = exports.MappedType = exports.LiteralType = exports.IntrinsicType = exports.IntersectionType = exports.InferredType = exports.IndexedAccessType = exports.ConditionalType = exports.ArrayType = exports.Type = void 0;
var abstract_1 = require("./abstract");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return abstract_1.Type; } });
var array_1 = require("./array");
Object.defineProperty(exports, "ArrayType", { enumerable: true, get: function () { return array_1.ArrayType; } });
var conditional_1 = require("./conditional");
Object.defineProperty(exports, "ConditionalType", { enumerable: true, get: function () { return conditional_1.ConditionalType; } });
var indexed_access_1 = require("./indexed-access");
Object.defineProperty(exports, "IndexedAccessType", { enumerable: true, get: function () { return indexed_access_1.IndexedAccessType; } });
var inferred_1 = require("./inferred");
Object.defineProperty(exports, "InferredType", { enumerable: true, get: function () { return inferred_1.InferredType; } });
var intersection_1 = require("./intersection");
Object.defineProperty(exports, "IntersectionType", { enumerable: true, get: function () { return intersection_1.IntersectionType; } });
var intrinsic_1 = require("./intrinsic");
Object.defineProperty(exports, "IntrinsicType", { enumerable: true, get: function () { return intrinsic_1.IntrinsicType; } });
var literal_1 = require("./literal");
Object.defineProperty(exports, "LiteralType", { enumerable: true, get: function () { return literal_1.LiteralType; } });
var mapped_1 = require("./mapped");
Object.defineProperty(exports, "MappedType", { enumerable: true, get: function () { return mapped_1.MappedType; } });
var optional_1 = require("./optional");
Object.defineProperty(exports, "OptionalType", { enumerable: true, get: function () { return optional_1.OptionalType; } });
var predicate_1 = require("./predicate");
Object.defineProperty(exports, "PredicateType", { enumerable: true, get: function () { return predicate_1.PredicateType; } });
var query_1 = require("./query");
Object.defineProperty(exports, "QueryType", { enumerable: true, get: function () { return query_1.QueryType; } });
var reference_1 = require("./reference");
Object.defineProperty(exports, "ReferenceType", { enumerable: true, get: function () { return reference_1.ReferenceType; } });
var reflection_1 = require("./reflection");
Object.defineProperty(exports, "ReflectionType", { enumerable: true, get: function () { return reflection_1.ReflectionType; } });
var rest_1 = require("./rest");
Object.defineProperty(exports, "RestType", { enumerable: true, get: function () { return rest_1.RestType; } });
var template_literal_1 = require("./template-literal");
Object.defineProperty(exports, "TemplateLiteralType", { enumerable: true, get: function () { return template_literal_1.TemplateLiteralType; } });
var tuple_1 = require("./tuple");
Object.defineProperty(exports, "NamedTupleMember", { enumerable: true, get: function () { return tuple_1.NamedTupleMember; } });
Object.defineProperty(exports, "TupleType", { enumerable: true, get: function () { return tuple_1.TupleType; } });
var type_operator_1 = require("./type-operator");
Object.defineProperty(exports, "TypeOperatorType", { enumerable: true, get: function () { return type_operator_1.TypeOperatorType; } });
var type_parameter_1 = require("./type-parameter");
Object.defineProperty(exports, "TypeParameterType", { enumerable: true, get: function () { return type_parameter_1.TypeParameterType; } });
var union_1 = require("./union");
Object.defineProperty(exports, "UnionType", { enumerable: true, get: function () { return union_1.UnionType; } });
var unknown_1 = require("./unknown");
Object.defineProperty(exports, "UnknownType", { enumerable: true, get: function () { return unknown_1.UnknownType; } });
