// typedoc/dist/lib/output/theme.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Theme = void 0;
const components_1 = require("./components");
const component_1 = require("../utils/component");
const resources_1 = require("./utils/resources");
/**
 * Base class of all themes.
 *
 * A theme defines the logical and graphical output of a documentation. Themes are
 * directories containing a ```theme.js``` file defining a [[BaseTheme]] subclass and a
 * series of subdirectories containing templates and assets. You can select a theme
 * through the ```--theme <path/to/theme>``` commandline argument.
 *
 * The theme class controls which files will be created through the [[BaseTheme.getUrls]]
 * function. It returns an array of [[UrlMapping]] instances defining the target files, models
 * and templates to use. Additionally themes can subscribe to the events emitted by
 * [[Renderer]] to control and manipulate the output process.
 *
 * The default file structure of a theme looks like this:
 *
 * - ```/assets/```<br>
 *   Contains static assets like stylesheets, images or javascript files used by the theme.
 *   The [[AssetsPlugin]] will deep copy this directory to the output directory.
 *
 * - ```/layouts/```<br>
 *   Contains layout templates that the [[LayoutPlugin]] wraps around the output of the
 *   page template. Currently only one ```default.hbs``` layout is supported. Layout templates
 *   receive the current [[PageEvent]] instance as their handlebars context. Place the
 *   ```{{{contents}}}``` variable to render the actual body of the document within this template.
 *
 * - ```/partials/```<br>
 *   Contains partial templates that can be used by other templates using handlebars partial
 *   syntax ```{{> partial-name}}```. The [[PartialsPlugin]] loads all files in this directory
 *   and combines them with the partials of the default theme.
 *
 * - ```/templates/```<br>
 *   Contains the main templates of the theme. The theme maps models to these templates through
 *   the [[BaseTheme.getUrls]] function. If the [[Renderer.getTemplate]] function cannot find a
 *   given template within this directory, it will try to find it in the default theme
 *   ```/templates/``` directory. Templates receive the current [[PageEvent]] instance as
 *   their handlebars context. You can access the target model through the ```{{model}}``` variable.
 *
 * - ```/theme.js```<br>
 *   A javascript file that returns the definition of a [[BaseTheme]] subclass. This file will
 *   be executed within the context of TypeDoc, one may directly access all classes and functions
 *   of TypeDoc. If this file is not present, an instance of [[DefaultTheme]] will be used to render
 *   this theme.
 */
let Theme = class Theme extends components_1.RendererComponent {
    /**
     * Create a new BaseTheme instance.
     *
     * @param renderer  The renderer this theme is attached to.
     * @param basePath  The base path of this theme.
     */
    constructor(renderer, basePath) {
        super(renderer);
        this.basePath = basePath;
        this.resources = new resources_1.Resources(this);
    }
};
Theme = __decorate([
    (0, component_1.Component)({ name: "theme", internal: true })
], Theme);
exports.Theme = Theme;
